app/services/duplicate_profile_checker.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+class DuplicateProfileChecker
+  attr_reader :user, :user_session, :sp
+
+  def initialize(user:, user_session:, sp:)
+    @user = user
+    @user_session = user_session
+    @sp = sp
+  end
+
+  def validate_user_does_not_have_duplicate_profile
+    return unless sp_eligible_for_one_account?
+    return unless user_has_ial2_profile?
+    return if user_profile_already_validated?
+    cacher = Pii::Cacher.new(user, user_session)
+    profile_id = user&.active_profile&.id
+    pii = cacher.fetch(profile_id)
+    duplicate_ssn_finder = Idv::DuplicateSsnFinder.new(user:, ssn: pii[:ssn])
+
+    if !duplicate_ssn_finder.ssn_is_unique?
+      DuplicateProfileConfirmation.create(
+        profile_id: profile_id,
+        confirmed_at: Time.zone.now,
+        duplicate_profile_ids: duplicate_ssn_finder.associated_profiles_with_matching_ssn.map(&:id),
+      )
+    end
+  end
+
+  private
+
+  def sp_eligible_for_one_account?
+    return false unless sp.present?
+    IdentityConfig.store.eligible_one_account_providers.include?(sp&.friendly_name)
+  end
+
+  def user_profile_already_validated?
+    DuplicateProfileConfirmation.where(profile_id: user&.active_profile&.id).present?
```

Comment:
We should be confirming that no new duplicate profiles have been added since the last check, if any.

Reply:
Are we expecting this to be a constant query, do we fear how expensive this could be for every login? May be better to think future state and check every month or year?

Reply:
If you are talking about this query `Profile.where.not(user_id: user.id).where(ssn_signature: ssn_signatures).empty?`, it should be performant as we have indexes on `ssn_signature`. We could resort to checking more infrequently if it does prove to be a performance issue.